File: sdk/swift/Sources/TokenPayID/TpidLoginButton.swift
import SwiftUI

/// Visual variants for ``TpidLoginButton``.
///
/// - `dark`     — black pill, white "TOKEN PAY ID" wordmark.
/// - `light`    — white pill with border, black wordmark.
/// - `iconOnly` — circular 1:1 button, just the T icon.
public enum TpidButtonVariant: Sendable { case dark, light, iconOnly }

/// Size presets matching the design system.
public enum TpidButtonSize: Sendable {
    case small, medium, large

    /// Overall button height / diameter.
    var height: CGFloat {
        switch self {
        case .small: return 40
        case .medium: return 48
        case .large: return 56
        }
    }
    /// Height of the wordmark image on DARK / LIGHT variants.
    /// In 2.5.0-pre.3 bumped from 16/20/24 to 18/22/26 pt — second-round
    /// user feedback ("чтобы лого был немного крупнее"). The pill heights
    /// grew proportionally so the wordmark still has breathing room.
    var wordmarkHeight: CGFloat {
        switch self {
        case .small: return 18
        case .medium: return 22
        case .large: return 26
        }
    }
    /// Size of the single-glyph icon on ICON_ONLY.
    var iconHeight: CGFloat {
        switch self {
        case .small: return 22
        case .medium: return 28
        case .large: return 34
        }
    }
    var padH: CGFloat {
        switch self {
        case .small: return 20
        case .medium: return 26
        case .large: return 32
        }
    }
}

/// Official TOKEN PAY ID login button — only the logo is rendered, so every
/// integration is visually identical. Wraps presentation of the widget sheet.
///
/// ```swift
/// TpidLoginButton(variant: .dark) { result in … }
/// ```
public struct TpidLoginButton: View {
    private let variant: TpidButtonVariant
    private let size: TpidButtonSize
    private let fullWidth: Bool
    private let onResult: @Sendable (TpidResult) -> Void
    @State private var presenting = false
    @State private var inFlight = false

    public init(
        variant: TpidButtonVariant = .dark,
        size: TpidButtonSize = .medium,
        fullWidth: Bool = false,
        onResult: @escaping @Sendable (TpidResult) -> Void
    ) {
        self.variant = variant
        self.size = size
        self.fullWidth = fullWidth
        self.onResult = onResult
    }

    public var body: some View {
        Button {
            presenting = true
            inFlight = true
        } label: {
            buttonBody
        }
        .buttonStyle(TpidLoginButtonStyle())
        .disabled(inFlight && !presenting)
        .sheet(isPresented: $presenting) {
            TpidWidgetSheet { result in
                presenting = false
                inFlight = false
                onResult(result)
            }
        }
    }

    @ViewBuilder
    private var buttonBody: some View {
        switch variant {
        case .dark, .light:
            let dark = variant == .dark
            let bg: Color = dark ? Color(red: 11/255, green: 11/255, blue: 13/255) : .white
            // Equal-weight border on both themes — fixes the feedback that the
            // black pill looked "bare" next to the outlined white one.
            // Pronounced outline — 2pt × 28% alpha reads as a clear ring
            // on both themes. Mirrors the Android 2.5.0-pre.3 bump.
            let borderCol: Color = dark
                ? Color.white.opacity(0.28)
                : Color(red: 11/255, green: 11/255, blue: 13/255).opacity(0.28)
            let shadowAlpha: Double = dark ? 0.50 : 0.18
            HStack(spacing: 0) {
                if inFlight {
                    ProgressView()
                        .controlSize(.small)
                        .tint(dark ? .white : .black)
                        .frame(height: size.wordmarkHeight)
                } else {
                    Image(dark ? "tpid_wordmark_white" : "tpid_wordmark_black", bundle: .module)
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .frame(height: size.wordmarkHeight)
                        .accessibilityLabel("Sign in with TOKEN PAY ID")
                }
            }
            .frame(maxWidth: fullWidth ? .infinity : nil, minHeight: size.height)
            .padding(.horizontal, size.padH)
            .background(Capsule().fill(bg))
            .overlay(Capsule().strokeBorder(borderCol, lineWidth: 2.0))
            .shadow(color: Color.black.opacity(shadowAlpha), radius: 6, x: 0, y: 2)
            .contentShape(Capsule())

        case .iconOnly:
            // White-tinted saturn glyph on a soft black disc, matching the
            // round brand badge in the tokenpay.space hero. A faint white
            // ring at the edge keeps it from looking like a flat disc.
            ZStack {
                Circle().fill(Color(red: 11/255, green: 11/255, blue: 13/255))
                if inFlight {
                    ProgressView().controlSize(.small).tint(.white)
                } else {
                    Image("tpid_icon", bundle: .module)
                        .resizable()
                        .renderingMode(.template)
                        .aspectRatio(contentMode: .fit)
                        .foregroundStyle(Color.white)
                        .frame(width: size.iconHeight, height: size.iconHeight)
                }
            }
            .frame(width: size.height, height: size.height)
            .overlay(Circle().strokeBorder(Color.white.opacity(0.28), lineWidth: 2.0))
            .shadow(color: Color.black.opacity(0.45), radius: 6, x: 0, y: 2)
            .contentShape(Circle())
        }
    }
}

/// Custom button style that scales down on press — mirrors the spring
/// scale(0.965 → 1.0) animation we use on JVM and Android so the three
/// native SDKs behave identically when tapped.
private struct TpidLoginButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.965 : 1.0)
            .animation(.spring(response: 0.26, dampingFraction: 0.6), value: configuration.isPressed)
    }
}

File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/TpidLoginButton.kt
package space.tokenpay.id

import android.app.Activity
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsHoveredAsState
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.hoverable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch

/**
 * TOKEN PAY ID login button. Three variants match the official design system:
 *
 * - [TpidButtonVariant.Dark]  — black pill, white "TOKEN PAY ID" wordmark.
 * - [TpidButtonVariant.Light] — white pill with border, black wordmark.
 * - [TpidButtonVariant.IconOnly] — circular button with the brand glyph.
 *
 * The button *only* shows the official logo — no custom label — so every
 * integration across apps and websites looks identical.
 */
@Composable
fun TpidLoginButton(
    modifier: Modifier = Modifier,
    variant: TpidButtonVariant = TpidButtonVariant.Dark,
    size: TpidButtonSize = TpidButtonSize.Medium,
    loading: Boolean = false,
    fullWidth: Boolean = false,
    onResult: (TpidResult) -> Unit,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    var inFlight by remember { mutableStateOf(false) }
    val activity = remember(context) {
        var c = context
        while (c is android.content.ContextWrapper && c !is Activity) c = c.baseContext
        c as? Activity
    }
    val enabled = !loading && !inFlight && activity != null

    val onClick: () -> Unit = click@{
        val act = activity ?: return@click
        inFlight = true
        scope.launch {
            val r = runCatching { TpidWidget.present(act) }
                .getOrElse { TpidResult.Failure(TpidError.Unknown(it)) }
            inFlight = false
            onResult(r)
        }
    }
    val dims = size.dims()

    // Shared hover/press interaction source drives the same scale + elevation
    // animation across all three variants so they feel identical in hand.
    val interaction = remember { MutableInteractionSource() }
    val hovered by interaction.collectIsHoveredAsState()
    val pressed by interaction.collectIsPressedAsState()
    val scale by animateFloatAsState(
        targetValue = if (pressed) 0.965f else if (hovered) 1.015f else 1.0f,
        animationSpec = spring(dampingRatio = 0.6f, stiffness = 520f),
        label = "tpid-btn-scale",
    )

    when (variant) {
        TpidButtonVariant.Dark, TpidButtonVariant.Light -> {
            val dark = variant == TpidButtonVariant.Dark
            val bg = if (dark) Color(0xFF0B0B0D) else Color.White
            val fg = if (dark) Color.White else Color(0xFF0B0B0D)
            // Equal-weight border on both themes — fixes the user's "чёрная и
            // белая кнопки пока не очень классные" feedback where the white
            // pill looked outlined and the black one looked bare.
            // Pronounced outline — 2dp × 28% alpha reads as a clear ring
            // on both themes. Previously 1.5dp × 18% alpha, which the user
            // said "не хватает окантовки" on the DARK pill.
            val borderCol = if (dark) Color.White.copy(alpha = 0.28f)
                            else Color(0xFF0B0B0D).copy(alpha = 0.28f)
            val wmRes = if (dark) R.drawable.tpid_wordmark_white else R.drawable.tpid_wordmark_black
            val shape = RoundedCornerShape(dims.corner)
            val shadowAlpha = if (dark) 0.5f else 0.18f
            Row(
                modifier = modifier
                    .then(if (fullWidth) Modifier.fillMaxWidth() else Modifier)
                    .heightIn(min = dims.height)
                    .scale(scale)
                    .shadow(
                        elevation = if (hovered) 10.dp else 4.dp,
                        shape = shape,
                        clip = false,
                        ambientColor = Color.Black.copy(alpha = shadowAlpha),
                        spotColor = Color.Black.copy(alpha = shadowAlpha),
                    )
                    .clip(shape)
                    .background(bg)
                    .border(BorderStroke(2.dp, borderCol), shape)
                    .hoverable(interaction)
                    .clickable(
                        interactionSource = interaction,
                        indication = null,
                        enabled = enabled,
                        onClick = onClick,
                    )
                    .padding(horizontal = dims.padH, vertical = dims.padV),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (inFlight) {
                    CircularProgressIndicator(
                        strokeWidth = 2.dp, color = fg,
                        modifier = Modifier.size(dims.wordmark),
                    )
                } else {
                    Image(
                        painter = painterResource(id = wmRes),
                        contentDescription = "Sign in with TOKEN PAY ID",
                        modifier = Modifier.height(dims.wordmark),
                    )
                }
            }
        }
        TpidButtonVariant.IconOnly -> {
            // Circular black button with the brand glyph painted white so it
            // matches the saturn logo used in the tokenpay.space hero and
            // the wordmark on the DARK pill.
            Box(
                modifier = modifier
                    .size(dims.height)
                    .scale(scale)
                    .shadow(
                        elevation = if (hovered) 10.dp else 4.dp,
                        shape = CircleShape,
                        clip = false,
                        ambientColor = Color.Black.copy(alpha = 0.45f),
                        spotColor = Color.Black.copy(alpha = 0.45f),
                    )
                    .clip(CircleShape)
                    .background(Color(0xFF0B0B0D))
                    .border(BorderStroke(2.dp, Color.White.copy(alpha = 0.28f)), CircleShape)
                    .hoverable(interaction)
                    .clickable(
                        interactionSource = interaction,
                        indication = null,
                        enabled = enabled,
                        onClick = onClick,
                    ),
                contentAlignment = Alignment.Center,
            ) {
                if (inFlight) {
                    CircularProgressIndicator(
                        strokeWidth = 2.dp, color = Color.White,
                        modifier = Modifier.size(dims.icon),
                    )
                } else {
                    Image(
                        painter = painterResource(id = R.drawable.tpid_icon),
                        contentDescription = "Sign in with TOKEN PAY ID",
                        modifier = Modifier.size(dims.icon),
                        colorFilter = ColorFilter.tint(Color.White),
                    )
                }
            }
        }
    }
}

enum class TpidButtonVariant { Dark, Light, IconOnly }

enum class TpidButtonSize { Small, Medium, Large }

/**
 * [wordmark] — height of the wordmark image on DARK / LIGHT variants.
 * [icon]     — size of the single-glyph icon on ICON_ONLY.
 */
private data class Dims(
    val height: Dp, val wordmark: Dp, val icon: Dp,
    val padH: Dp, val padV: Dp, val corner: Dp,
)

private fun TpidButtonSize.dims(): Dims = when (this) {
    // Wordmark sized 18/22/26 dp on Small/Medium/Large after a second round
    // of user feedback ("чтобы лого был немного крупнее"). The pill heights
    // and horizontal padding grow a touch so the logo still has breathing
    // room inside the rounded capsule.
    TpidButtonSize.Small  -> Dims(40.dp, 18.dp, 22.dp, 20.dp, 9.dp, 999.dp)
    TpidButtonSize.Medium -> Dims(48.dp, 22.dp, 28.dp, 26.dp, 11.dp, 999.dp)
    TpidButtonSize.Large  -> Dims(56.dp, 26.dp, 34.dp, 32.dp, 13.dp, 999.dp)
}
